### PR TITLE
Add Stable Diffusion XL inference support

### LINKED ISSUE #148: Add Support for SDXL
Please add support for compiling stabilityai/stable-diffusion-xl-base-0.9 for inf2.

### PR DESCRIPTION
**Pull request by JingyaHuang**
- [x] Fix unet inference issue
- [x] Flag for accelerating the compilation
- [x] SDXL pipeline for inference
- [x] SDXL tests
- [x] SDXL documentation

fixes #148

### CODE REVIEW

File: docs/source/guides/models.mdx [hunk 1]
```diff
@@ -247,7 +247,40 @@ Now generate an image with a prompt on neuron:
 
 <img
   src="https://raw.githubusercontent.com/huggingface/optimum-neuron/main/docs/assets/guides/models/01-sd-image.png"
-  alt="search ami"
+  alt="stable diffusion generated image"
+/>
+
+## Stable Diffusion XL
+
+Similar to the stable diffusion, you will be able to use `NeuronStableDiffusionXLPipeline` API to export and run inference on Neuron devices with SDXL models.
```
Comment (pcuenca) [inline docs/source/guides/models.mdx:255]:
```suggestion
Similar to Stable Diffusion, you will be able to use `NeuronStableDiffusionXLPipeline` API to export and run inference on Neuron devices with SDXL models. 
```

**Review by pcuenca — COMMENTED** [review]
Looks great 🙌 , although I haven't tested it yet. I wonder if we can set `--enable-fast-loading-neuron-binaries` for `text_encoder_2` as well.

Are there plans to support the refiner phase? (just curious)

File: optimum/exporters/neuron/convert.py [hunk 2]
```diff
@@ -454,11 +460,14 @@ def export_neuronx(
 
 def add_stable_diffusion_compiler_args(config, compiler_args):
     if hasattr(config._config, "_name_or_path"):
-        sd_components = ["text_encoder", "unet", "vae", "vae_encoder", "vae_decoder"]
+        sd_components = ["text_encoder", "vae", "vae_encoder", "vae_decoder"]
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:463]: Good catch! Will add fast-loading for text encoder 2 as well.

**Comment by JingyaHuang** [comment]
Thanks for the review @pcuenca! The refiner support is planned for a coming PR.

File: optimum/exporters/neuron/convert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply optimum/exporters/neuron/convert.py:463]: Oh yes, you are right, sorry!

File: docs/assets/guides/models/02-sdxl-image.png
Comment (philschmid) [inline docs/assets/guides/models/02-sdxl-image.png:1]: 02-sdxl-image
here is a 10x smaller compressed version

File: docs/source/guides/models.mdx [hunk 3]
```diff
@@ -247,7 +247,40 @@ Now generate an image with a prompt on neuron:
...
+## Stable Diffusion XL
+
+Similar to Stable Diffusion, you will be able to use `NeuronStableDiffusionXLPipeline` API to export and run inference on Neuron devices with SDXL models. 
+
+```python
+>>> from optimum.neuron import NeuronStableDiffusionXLPipeline
+
+>>> model_id = "stabilityai/stable-diffusion-xl-base-1.0"
+>>> compiler_args = {"auto_cast": "matmul", "auto_cast_type": "bf16"}
+>>> input_shapes = {"batch_size": 1, "height": 1024, "width": 1024}
```
Comment (philschmid) [inline docs/source/guides/models.mdx:262]: is batch_size <> num_images_per_prompt?

File: docs/source/guides/models.mdx [hunk 4]
```diff
@@ -247,7 +247,40 @@ Now generate an image with a prompt on neuron:
...
+
+Now generate an image with a prompt on neuron:
+
+```python
+>>> prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
+>>> image = stable_diffusion_xl(prompt).images[0]
+```
+
+<img
+  src="https://raw.githubusercontent.com/huggingface/optimum-neuron/main/docs/assets/guides/models/02-sdxl-image.png"
```
Comment (philschmid) [inline docs/source/guides/models.mdx:282]: change here when image changed

File: optimum/exporters/neuron/convert.py [hunk 5]
```diff
@@ -434,6 +438,8 @@ def export_neuronx(
         compiler_args.extend(["--auto-cast-type", auto_cast_type])
     else:
         compiler_args = ["--auto-cast", "none"]
+    # WARNING: Enabled experimental parallel compilation
+    compiler_args.extend(["--enable-experimental-O1"])
```
Comment (philschmid) [inline optimum/exporters/neuron/convert.py:442]: should we add `f"--num-parallel-jobs {os.cpu_count()}"` to maximize the compilation? 

https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/compiler/neuronx-cc/index.html?highlight=enable-experimental-O1#id1

File: optimum/exporters/neuron/utils.py [hunk 6]
```diff
@@ -92,22 +90,23 @@ def build_stable_diffusion_components_mandatory_shapes(
     vae_decoder_num_channels: Optional[int] = None,
     height: Optional[int] = None,
     width: Optional[int] = None,
+    num_images_per_prompt: Optional[int] = 1,
```
Comment (philschmid) [inline optimum/exporters/neuron/utils.py:93]: are we storing the value in the config?

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 7]
```diff
@@ -189,6 +189,12 @@ def __call__(
         # 0. Height and width to unet (static shapes)
         height = self.unet.config.neuron["static_height"] * self.vae_scale_factor
         width = self.unet.config.neuron["static_width"] * self.vae_scale_factor
+        if self.num_images_per_prompt != num_images_per_prompt and not self.dynamic_batch_size:
```
Comment (philschmid) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion.py:192]: 👍🏻

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion_xl.py [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+
+        Returns:
+            [`~pipelines.stable_diffusion_xl.StableDiffusionXLPipelineOutput`] or `tuple`:
+            [`~pipelines.stable_diffusion_xl.StableDiffusionXLPipelineOutput`] if `return_dict` is True, otherwise a
+            `tuple`. When returning a tuple, the first element is a list with the generated images.
+        """
+        # 0. Default height and width to unet (static shapes)
+        height = self.unet.config.neuron["static_height"] * self.vae_scale_factor
+        width = self.unet.config.neuron["static_width"] * self.vae_scale_factor
+        if self.num_images_per_prompt != num_images_per_prompt and not self.dynamic_batch_size:
```
Comment (philschmid) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion_xl.py:366]: 👍🏻

File: tests/inference/test_stable_diffusion_pipeline.py [hunk 9]
```diff
@@ -89,3 +84,73 @@ def test_export_and_inference_dyn(self, model_arch):
...
         self.assertIsInstance(image, PIL.Image.Image)
+
+
+@is_inferentia_test
+@requires_neuronx
+@require_diffusers
+class NeuronStableDiffusionXLPipelineIntegrationTest(unittest.TestCase):
+    NEURON_MODEL_CLASS = NeuronStableDiffusionXLPipeline
+    STATIC_INPUTS_SHAPES = {"batch_size": 1, "height": 64, "width": 64}
+    COMPILER_ARGS = {"auto_cast": "matmul", "auto_cast_type": "bf16"}
```
Comment (philschmid) [inline tests/inference/test_stable_diffusion_pipeline.py:95]:
```suggestion
    COMPILER_ARGS = {"auto_cast": "all", "auto_cast_type": "bf16"}
```

**Review by philschmid — APPROVED** [review]
LGTM! 🔥 will test again! Added some question and add a compressed image, see comment.

File: optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion_xl.py [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+        negative_prompt_embeds: Optional[torch.FloatTensor] = None,
+        pooled_prompt_embeds: Optional[torch.FloatTensor] = None,
+        negative_pooled_prompt_embeds: Optional[torch.FloatTensor] = None,
+        lora_scale: Optional[float] = None,
+    ):
+        r"""
+        Encodes the prompt into text encoder hidden states.
+
+        Args:
+            prompt (`str`, *optional*):
```
Comment (michaelbenayoun) [inline optimum/neuron/pipelines/diffusers/pipeline_stable_diffusion_xl.py:49]:
```suggestion
            prompt (`Optional[str]`, defaults to `None`):
```

File: docs/source/guides/models.mdx [hunk 11]
```diff
@@ -247,7 +247,40 @@ Now generate an image with a prompt on neuron:
...
+>>> input_shapes = {"batch_size": 1, "height": 1024, "width": 1024}
+
+>>> stable_diffusion_xl = NeuronStableDiffusionXLPipeline.from_pretrained(model_id, export=True, **compiler_args, **input_shapes)
+
+# Save locally or upload to the HuggingFace Hub
+>>> save_directory = "sd_neuron_xl/"
+>>> stable_diffusion_xl.save_pretrained(save_directory)
+>>> stable_diffusion_xl.push_to_hub(
+...     save_directory, repository_id="my-neuron-repo", use_auth_token=True
+... )
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply docs/source/guides/models.mdx:271]: Yeah, you are right. We can do some refactoring to gather those functions in a `XXXMixin`.

[the remark this answers is not in the document]
Comment (JingyaHuang) [reply docs/source/guides/models.mdx:271]: Do you want to open a PR? or I can open one as well.

File: optimum/neuron/modeling_diffusion.py [hunk 12]
```diff
@@ -515,5 +624,63 @@ def forward(self, latent_sample: torch.Tensor):
 
 
 class NeuronStableDiffusionPipeline(NeuronStableDiffusionPipelineBase, StableDiffusionPipelineMixin):
-    def __call__(self, *args, **kwargs):
-        return StableDiffusionPipelineMixin.__call__(self, *args, **kwargs)
+    __call__ = StableDiffusionPipelineMixin.__call__
+
+
+class NeuronStableDiffusionXLPipelineBase(NeuronStableDiffusionPipelineBase):
+    # `TasksManager` registered img2ime pipeline for `stable-diffusion-xl`: https://github.com/huggingface/optimum/blob/v1.12.0/optimum/exporters/tasks.py#L174
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/neuron/modeling_diffusion.py:631]: Yeah, as I just need it to infer the task, but only `StableDiffusionXLImg2ImgPipeline` is registered in the TasksManager so far.
I need to make a PR to optimum main, but just want to get the PR in optimum-neuron first.

File: docs/assets/guides/models/02-sdxl-image.png
Comment (JingyaHuang) [reply docs/assets/guides/models/02-sdxl-image.png:1]: Will use it in the doc, thx!

File: docs/source/guides/models.mdx (hunk 3 above)
Comment (JingyaHuang) [reply docs/source/guides/models.mdx:262]: here the batch size indicate the batch size of prompt, `num_images_per_prompt` is default to 1 if not given.

File: optimum/exporters/neuron/utils.py (hunk 6 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/utils.py:93]: No, in the end, we register the actual `batch_size` (`batch_size x num_images_per_prompt`) and infer directly the `num_images_per_prompt` by `batch_size` used for compiling unet or vae decoder divided by `batch_size` used for compiling text encoder or vae encoder. 

The reason is that I don't want `num_images_per_prompt` to increase the code complexity if it doesn't hurt user experience.

File: optimum/exporters/neuron/convert.py (hunk 5 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:441]: I did not see any improvement on the compilation time with the flag neither.

**Review by dacorvo — COMMENTED** [review]
I don't know enough SDXL to do a very detailed review, but that is an impressive peace of work (and the images are stunning !).
LGTM.

**Pull request merged**